src/graphql/object_type.cr

```diff
@@ -172,6 +82,153 @@ module GraphQL::ObjectType
+    end
+
+    if inc = selection.directives.find { |d| d.name == "include" }
+      if arg = inc.arguments.find { |a| a.name == "if" }
+        return true if !arg.value.as(Bool)
+      end
+    end
+
+    false
+  end
+
+  # :nodoc:
+  protected def _graphql_resolve(context, selections : Array(::GraphQL::Language::Selection), json : JSON::Builder) : Array(::GraphQL::Error)
+    errors = [] of ::GraphQL::Error
+    json_fragments = Hash(String, Channel(JSONFragment)).new
+
+    selections.each do |selection|
+      case selection
+      when ::GraphQL::Language::Field
+        next if _graphql_skip?(selection)
+        path = selection._alias || selection.name
+        json_fragments[path] = Channel(JSONFragment).new
+
+        spawn do
+          fragment = _graphql_build_json_fragment(context, path) do |json|
+            _graphql_resolve(context, selection, json)
+          end
+
+          json_fragments[path].send fragment
+        end
+      when ::GraphQL::Language::FragmentSpread
+        next if _graphql_skip?(selection)
+
+        begin
+          errors.concat _graphql_resolve(context, selection, json)
+        rescue e
+          if message = context.handle_exception(e)
+            errors << ::GraphQL::Error.new(message, selection.name)
+          end
+        end
+      when ::GraphQL::Language::InlineFragment
+        next if _graphql_skip?(selection)
+
+        errors.concat _graphql_resolve(context, selection.selections, json)
+      else
+        # this never happens, only required due to Selection being turned into ASTNode
+        # https://crystal-lang.org/reference/1.3/syntax_and_semantics/virtual_and_abstract_types.html
+        raise ::GraphQL::TypeError.new("invalid selection type")
+      end
+    end
+
+    json_fragments.each do |path, channel|
+      fragment = channel.receive
+      errors.concat fragment.errors
+      next if fragment.json.empty?
+
+      json.field(path) { json.raw fragment.json }
+    end
+
+    errors
+  end
+
+  # :nodoc:
+  private def _graphql_resolve(context, fragment : ::GraphQL::Language::FragmentSpread, json : JSON::Builder) : Array(::GraphQL::Error)
+    errors = [] of ::GraphQL::Error
+    f = context.fragments.find{ |f| f.name == fragment.name}
+    if f.nil?
+      errors << ::GraphQL::Error.new("no fragment #{fragment.name}", fragment.name)
+    else
+      errors.concat _graphql_resolve(context, f.selections, json)
+    end
+    errors
+  end
+
+  # :nodoc:
+  private def _graphql_build_json_fragment(context, path : String | Array(Int32 | String), &block : JSON::Builder -> Array(::GraphQL::Error)) : JSONFragment
+    errors = [] of ::GraphQL::Error
+
+    json = String.build do |io|
+      builder = JSON::Builder.new(io)
+      builder.document do
+        errors.concat yield builder
+      end
+    rescue e
+      if message = context.handle_exception(e)
+        errors << ::GraphQL::Error.new(message, path)
+      end
+    end
+
+    JSONFragment.new(json, errors)
+  end
```

Comment:
I moved these outside of the included macro since they're plain ol' crystal methods so no need to redefine them on every including type.

Reply:
The macro hooks only affect when macros are executed; all methods in `ObjectType` will be included in objects because that's what [include](https://crystal-lang.org/reference/1.6/syntax_and_semantics/modules.html) does. So you'd need to move them elsewhere, but I think the `_graphql_resolve` methods should remain here at least.

Reply:
You're correct of course that they're included in the including type, but they're not _declared_ in the including type. Stack traces for example would contain the method in GraphQL::ObjectType and not the including class.

Reply:
Makes sense :shipit: